### ISSUE TITLE
End2end tests findings

### ISSUE THREAD

**Issue by berndbohmeier** [trustlines-protocol/blockchain#358]
I will just write down some of my findings, to discuss if we can improve them and then create tasks to do so. 

- We will not do this. I did not catch the problem that I have to wait for the reorg depth. One of our goals was to make the bridge more talk to find problems more easy, so maybe we should log every event already on "latest" or even "pending" with the note that we will not react to it now, but wait for the reorg depth. 

- Deferred ~The confirmations of the bridge were failing because the bridge had no funds, but there was no logging of the bridge that the transaction were failing. I dont know the exact reason why this is not logged, as I can see some code that should do this I think? We should make sure that this is logged.~ This was also because of the reorg depth. Same applies as above maybe we want to log it already before? I dont know. 

- [x] We could also check that the bridge has enough funds? But we probably do not need it as we will fund the bridge only once and this is more a problem that arises in testing. 

- [x] Some of the services use the block time as poll interval, we should make those poll intervals also configurable. We can of course use the block time still as the default. 

- [x] To make the tests more reliable it would be good to have a way to test if the bridge is "up" whatever that means, maybe you have an idea how to do that.

**Comment by schmir-at-bb** [comment]
> I will just write down some of my findings, to discuss if we can improve them and then create tasks to do so.
> 
>     * I did not catch the problem that I have to wait for the reorg depth. One of our goals was to make the bridge more talk to find problems more easy, so maybe we should log every event already on "latest" or even "pending" with the note that we will not react to it now, but wait for the reorg depth.

We don't even fetch them at the moment. I'm not sure if it makes sense, I would rather not do this at the moment. 

If this becomes a problem in a real world setup, it's because the chain doesn't produce new blocks or we don't receive any. This is something I would rather try to determine and log.

> 
>     * ~The confirmations of the bridge were failing because the bridge had no funds, but there was no logging of the bridge that the transaction were failing. I dont know the exact reason why this is not logged, as I can see some code that should do this I think? We should make sure that this is logged.~ This was also because of the reorg depth. Same applies as above maybe we want to log it already before? I dont know.

I would log the fact that we're waiting for more blocks if we see a transaction that's not yet reorg safe. We have that information already (at least for the first transaction that we're waiting for)

> 
>     * We could also check that the bridge has enough funds? But we probably do not need it as we will fund the bridge only once and this is more a problem that arises in testing.

Then let's at least abort the program if the bridge has zero funds when starting tlbc-bridge.

> 
>     * Some of the services use the block time as poll interval, we should make those poll intervals also configurable. We can of course use the block time still as the default.

ok. I would then actually also make the HOME_CHAIN_STEP_DURATION that's currently living in bridge.constants configurable.

> 
>     * To make the tests more reliable it would be good to have a way to test if the bridge is "up" whatever that means, maybe you have an idea how to do that.

It has determined it's a validator and it has enough funds (i.e. it's validating).
Maybe that we have seen events from both chains (including the FetcherReachedHead event, when we need to wait for new blocks).

But, how would we report that back? The end2end tests could either parse the logs or we could start an optional HTTP server that provides some basic information about the current state of the system (as json).

The latter solution would be useful even when not running the tests.

**Comment by berndbohmeier** [comment]
I agree with all the things you said. 

If we decide that we want to have a small http server to ask for the status, this would be awesome. 
Also easier for the tests, as I do not really like having to parse the logs. Those are human readable messages and maybe not that stable. 
But if we decide the server is too much for now, we can also go for the log parsing.

**Comment by schmir-at-bb** [comment]
#369 is a first shot at the webservice. We should add some statistics about events handled, confirmations received, amount transferred...It's not there yet at the moment.

**Comment by schmir-at-bb** [comment]
#369 is merged. You have to enable it with

```
[webservice]
enabled = true
host = "127.0.0.1"
port = 9500
```
in the config file. Environment variables should still work (use something like `webservice.host`, though I haven't tested that.

The validator's key now has to be specified with
```
[validator_private_key]
raw = "0x7..."
```

**Comment by berndbohmeier** [comment]
Everything is done, or decided that we dont want to do this now.
